ament_cmake_pclint/cmake/ament_pclint.cmake

```diff
@@ -0,0 +1,145 @@
+# Copyright 2017 Apex.AI, Inc.
+#
+#
+# This file contains modified code from the following open source projects
+# published under the licenses listed below:
+#
+#
+# # Copyright 2014-2015 Open Source Robotics Foundation, Inc.
+# #
+# # Licensed under the Apache License, Version 2.0 (the "License");
+# # you may not use this file except in compliance with the License.
+# # You may obtain a copy of the License at
+# #
+# #     http://www.apache.org/licenses/LICENSE-2.0
+# #
+# # Unless required by applicable law or agreed to in writing, software
+# # distributed under the License is distributed on an "AS IS" BASIS,
+# # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# # See the License for the specific language governing permissions and
+# # limitations under the License.
+
+function(_parse_pclint_input)
+
+  find_program(ament_pclint_BIN NAMES "ament_pclint")
+  if(NOT ament_pclint_BIN)
+    message(FATAL_ERROR "ament_pclint() could not find program 'ament_pclint'")
+  endif()
+
+  set(cmd "${ament_pclint_BIN}" ${ARG_UNPARSED_ARGUMENTS})
+
+  get_property(INCLUDED_DIRS DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
```

Comment:
How does it handle the case that no global include directories are being used but target specific include directories instead?

Same for compiler definitions below.

Reply:
I'm not 100% I understand the question, but I think what I meant to do here was gather all the directories included through cmake's `include_directories`. The idea is that you need to tell pclint where to look for header files, not unlike how you have to tell cmake where to find header files. 

It's likely that I'm misusing or misunderstanding something in cmake, though.

Reply:
Similarly, we want to give pclint compiler definitions added through target_compile_definitions, but again, I may be misunderstanding something fundamental about cmake.

Reply:
The CMake code might set these information on a per-targetg level using e.g. `target_include_directories` and/or `target_compile_definitions`. In that case the global variables might not be sufficient.

Reply:
I see what you're saying now. I was assuming that any headers are included with `include_directories`

What I'm thinking is we need to add PCLINT_INCLUDE_DIRS and PCLINT_COMPILE_DEFS options to `ament_pclint` or do you know of something better to get information added via `target_*()`? 

Basically, pclint needs access to the exact same directories as the `add_executable` or `add_library` functions.

Reply:
I don't see a better way then to let the user pass in the necessary flags manually. That can be either the include dirs directory or a list of targets from which the include dirs are being extracted.

Beside supporting these information as arguments in the CMake function the user also needs to be able to set some CMake variables for the case where the linter is being invoked implicitly by the `ament_lint_auto` package.

Reply:
I was considering how to do this as well.  @jpsamper2009 is there not a way to retrieve all of the includes and flags for a particular target rather than the global scope in CMake?

Reply:
I think what you want to do is:

add_library(
  ${PROJECT_NAME}
  ${rcutils_sources})

```
get_target_property(TARGET_INC ${PROJECT_NAME} INCLUDE_DIRECTORIES)
message(STATUS "INCLUDES: " ${TARGET_INC})

get_target_property(TARGET_FLAGS ${PROJECT_NAME} COMPILE_FLAGS)
message(STATUS "FLAGS: " ${TARGET_FLAGS})

get_target_property(TARGET_DEF ${PROJECT_NAME} COMPILE_DEFINITIONS)
message(STATUS "DEFINITIONS: " ${TARGET_DEF})
```

And then use those generated variables to pass into pclint.

Reply:
Upon further reflection, I think that the appropriate way to handle `ament_cmake_pclint` is to add a CMake function that takes a target as an argument.  From there, we can retrieve the list of sources, includes, and compiler definitions/flags that are used on that target, to make the linting maximally effective.

This would also allow us to exclude things like unit tests, etc from pclint checking.

Reply:
@mjcarroll I hadn't seen your comment above, so for now I've just added function arguments to allow the user to specify the include directories explicitly.

Do you guys have an example of a package with target-specific include directories and compiler flags?